lib/command.ts

```diff
@@ -33,19 +33,59 @@ export default abstract class BalenaCommand extends Command {
 	 */
 	public static root = false;
 
-	protected async checkElevatedPrivileges() {
-		const root = (this.constructor as typeof BalenaCommand).root;
-		if (root) {
-			const isElevated = await (await import('is-elevated'))();
-			if (!isElevated) {
-				throw new ExpectedError(
-					'You need admin privileges to run this command',
-				);
-			}
+	/**
+	 * Require authentication to run.
+	 * When set to true, command will exit with an error
+	 * if user is not already logged in.
+	 */
+	public static authenticated = false;
+
+	/**
+	 * Throw InsufficientPrivilegesError if not root on Mac/Linux
+	 * or non-Administrator on Windows.
+	 *
+	 * Called automatically if `root=true`.
+	 * Can be called explicitly by command implementation, if e.g.:
+	 *  - check should only be done conditionally
+	 *  - other code needs to execute before check
+	 */
+	protected static async checkElevatedPrivileges() {
+		const isElevated = await (await import('is-elevated'))();
+		if (!isElevated) {
+			throw new InsufficientPrivilegesError(
+				'You need root/admin privileges to run this command',
+			);
 		}
 	}
 
+	/**
+	 * Throw NotLoggedInError if not logged in.
+	 *
+	 * Called automatically if `authenticated=true`.
+	 * Can be called explicitly by command implementation, if e.g.:
+	 *  - check should only be done conditionally
+	 *  - other code needs to execute before check
+	 *
+	 *  Note, currently public to allow use outside of derived commands
+	 *  (as some command implementations require this. Can be made protected
+	 *  if this changes).
+	 */
+	public static async checkLoggedIn() {
+		await (await import('./utils/patterns')).checkLoggedIn();
+	}
+
 	protected async init() {
-		await this.checkElevatedPrivileges();
+		const requireElevatedPrivileges = (this.constructor as typeof BalenaCommand)
+			.root;
+		const requireAuthentication = (this.constructor as typeof BalenaCommand)
+			.authenticated;
```

Comment:
@pdcastro 

>What I am reading in a web search is that overriding static properties / methods (and expecting the parent class to transparently use what the subclass overrode) is a bit of an OO anti-pattern -- if overriding is desirable, it should not be static. 

My main goal here was to have the custom properties behave in exactly the same way as the standard ones.  If these were not static then we'd have to treat them differently in the consuming commands, which I think would be confusing and inconsistent. oclif itself does the same thing with the debug property. (I think it's possible to find all sorts of opinions on the web about what's good-practice/anti-pattern :-)  BTW This isn't new to this PR, it's already in master since a PR last week.

> It feels like there should be a way of using Typescript's object model without needing this type cast.

I've simply copied oclifs own approach here.  [https://github.com/oclif/command/blob/master/src/command.ts#L87-L98]   (I copied their approach after I couldn't figure out how to do it after an hour or so).  

> If root, authenticated, checkElevatedPrivileges and checkLoggedIn properties / methods were non-static, then here the type cast could be avoided

Regarding `root`, `authenticated`.  IMO there's little benefit in that, and it comes with the downside of these custom properties having to be treated differently to the standard oclif  ones in the subclassed commands.

Regarding `checkElevatedPrivileges`, `checkLoggedIn`,  These were originally not static, but I made them so on the suggestion of @CameronDiver in a previous comment - there are some existing oclif commands that use functions in the command's module but not in the command's class, this means that member functions of the command can't be consumed in these functions, but by making them static they can be (these should probably be moved into their classes at some point).

Reply:
> By the way, if the subclass does not define the static root and authenticated properties, would these properties end up with the undefined value in this init method, and the tsc compiler silent about it? (I have not tested it, but if this.constructor is the subclass and the property is static, then maybe the subclass would have to override the properties, but without the compiler enforcing it.)

I've tested this now, and conform that if the properties are not defined in the subclass then it returns the values from the parent class.

Reply:
> there are some existing oclif commands that use functions in the command's module but not in the command's class, this means that member functions of the command can't be consumed in these functions

I think that's the case of the `validateOptions` example I gave above, where an alternative to using static methods is to pass an instance of the command as an argument to the function. But it's OK, mainly I wanted to avoid the type cast.

> I've tested this now, and confirm that if the properties are not defined in the subclass then it returns the values from the parent class.

Thanks for checking it. 👍  I am approving the PR.

Reply:
Ok cool, thanks.